### ISSUE TITLE
[core] RFC: draft for rule reorganisation

### ISSUE THREAD

**Issue by oowekyala** [pmd/pmd#610]
# Draft: rule reorganisation and simpler rule references

What this draft does address:
* Rule reorganisation
* Ruleset agnostic rule references, like what was proposed in the [thread linked](https://sourceforge.net/p/pmd/discussion/188194/thread/b840897c/) from #450

What this draft doesn't address:
* timeToFix attribute (see #329 )
* RulesetFactory refactoring
* @jsotuyod's idea about reworking messages in the XML
* Stricter validation constraints on the schema

## Long term goal
1. Rule names should be unique within a language module, exported rule references should look like "java/GodClassRule"
2. The "category" (or "categories") attribute should be mandatory for standard rules, stating at least one category. The first one stated will act as a "primary category", some category we're sure the rule will not change.
3. Other categories should act as tags for documentation, external tools, and exclusion from a ruleset.
4. Standard rules keep their definition in dedicated xml files, one for each primary category. 
5. We should generate an index of all rules for each language module at build time, and use that to resolve references of the form "lang/rulename" when building the user's ruleset.
6. Rulesets don't define rules anymore, they just use them. Our rulesets are thus considered like user ones, they correspond to a use case, not a rule definition unit.
7. Full ruleset references of the form "rulesets/java/basic.xml" and "java-basic" are kept in use. It's useful to import a ruleset all at once, even with exclusions.
8. References of the form "rulesets/java/basic.xml/BrokenNullCheck" are kept for backwards compatibility, but discouraged with a weak warning. Rulesets are no longer a reliable reference for individual rules, as they may change between releases. Use just "java/BrokenNullCheck"

### Additional thoughts
Numbers refer to the above list
1. I think today, it makes no real sense to enforce rule name unicity across all of pmd, given eg the vast number of Apex rules that were ported from Java. Besides, we strip details like rulesets from rule references to allow them to change. Rules cannot just move to another language.
2. We might disagree on the choice of the primary category, they do not correspond one to one to the current rulesets. For example "migrating" and its flavours should perhaps not be a primary category, more like secondaries/ rulesets.
3. I think, excluding a rule from a ruleset inclusion based on its category is the non-invasive and useful form of using categories in PMD as more than metadata. Full category inclusion would probably be difficult to implement and introduce inconsistencies. 
4. Rule definition files could perhaps use another schema to decouple theirs from the rulesets' one, though we can use the ruleset schema for now.
  It may not be a good idea to couple primary category with the files rules are defined in. I think that offers at least the advantage of organising the rules by topic, allowing us to spot categories which lack some rules easily. The structure would be flexible so that's not a dead-end. What do you think?
5. The index generator could be as simple as copying every rule definition xml element to a single xml file. We then validate that file for name collisions at test time.
6. Rulesets could be handy to provide default configs, as several ones could use the same rules

## Features that should be kept for 6.0.0
As time is ticking, there are some things that cannot make it through for 6.0.0. This is the minimum I think we should do to make any impact and have our hands free for further changes thereafter:
* Category attribute and rule reorganisation around primary category. Current rulesets still exist so that references are not broken.
* Rule references of the form "lang/rulename".

## Features that can be postponed
* Full refactoring of RulesetFactory. It would be really nice, but unless we draft this now too, I don't think we should bother with that
* Using secondary categories for documentation and ruleset exclusion is a non-breaking change, we could make that later
* Definitive structure for the primary categories. If we decouple rule definition from the rulesets, we can make any change to the rule definition structure later. That means we can use a temporary structure for categories (like simply using the ruleset name as primary) to set up the infrastructure and take that work off our plate
* Specific schema for rule definition files. That actually depends on whether we want users to have access to the category attribute or not, since it would be all but decorative for them. I think we may not care

## Implementation plan for 6.0.0

This is based on the "bare minimum scenario" exposed above.

Create category attribute:
* Add category attribute to rule nodes in the ruleset schema
* (Add "getCategories()" to the Rule interface, populate it in ruleset factory. That would only be useful for category exclusion in ruleset reference, and documentation, so maybe not today)

Foreach language:
* Create rule definition files:
  * Make a copy of ruleset files under src/main/resources/ruledef/langname. These will be our temporary primary categories
  * Populate their category attribute with the name of the ruleset with some Perl script or whatever 
  * The files under src/main/resources/rulesets stay the same until the new rule references work.
* Implement index generator:
   * Make a Java executable that copies all rule definition elements found in the files under resources/ruledef/langname/*.xml into a single file, like "resources/rule-index/langName/all-ruledefs.xml"
   * Unit test it
   * Plug it into the maven build using maven exec plugin, during the compile phase I guess

Make ruleset factory handle the new references:
* Create new class TerseRuleReferenceId or whatever so as not to touch any other class. Instances of this class identify a rule within a language, they can give the name of the rule, the language it lives in, and the overriden properties and attributes. However, it does not contain the rule to which we delegate, like RuleReference does. That's because we want to load the big rule index only once and get all the rules we need at the end, not load each rule at a time when seeing a rule element.
* Create a builder "MassRulesBuilder". Each time we find a rule element with a reference of the form "java/rulename", we add the TerseRuleReferenceId to the builder.
  * The "build" method returns a set of rules
  * Internally, we group all TerseRuleReferenceIds by the language module they refer to. The build() method could look like that in pseudocode:
```
// Map<String, List<TerseRuleReference>> terseRuleReferencesByLanguage
//...
foreach entry (languageName, list of TerseRuleReferenceIds) in the map:
    if this entry references at least one rule:
        get a reader for  resource "/rule-index/${languageName}/all-ruledefs.xml"
        
        Get the rule definition elements we want using the reader and the list of TerseRuleReferenceIds
        Instantiate the rules
        Override each rule's behaviour based on the elements overriden in the rset, which are stored in the TerseRuleReference 
        return the overriden Rules;
```
* We have a massRulesBuilder as instance variable of rulesetFactory, and call "build()" before returning the final ruleset
* Test that (I think that may be it, I surely must have missed something)

Finally:
  * Change the ruleset files to use the new rule references
  * Log a warning when references of the form "rulesets/java/basic.xml/BrokenNullCheck" are used
  * Documentation
  
### Discussion

* Do you agree with the long term vision?
* Do you think the implementation plan is realistic? 
* Can we begin implementing straight away? I'd be okay to start, I still have a lot of spare time this month

**Comment by oowekyala** [comment]
Update: I've begun working on that in [that branch](https://github.com/oowekyala/pmd/blob/short-references/pmd-core/src/main/java/net/sourceforge/pmd/RuleDefsAggregator.java). Currently the index generator seems done to me, but still needs some testing.

The files under `resources/ruledefs` are copies of the rulesets, I didn't bother to add the category attribute. It probably doesn't make sense while we keep this temporary structure.

One thing that occurred to me, is that this feature will need the rule doc generator to be updated into generating something else than only ruleset documentation. We can think about that later though

Lastly, I think RulesetFactory may be more affected by this feature than I had intended (for the best though). For instance, the logic behind references of the form `rulesets/java-basic.xml/Rule` would surely be worth refactoring. Eg, that reference could be transformed to `java/Rule` with regex, which means it can be converted to the new rule reference format easily, and the ruleset file need not even be read (provided the `java-basic` ruleset is not fully included elsewhere in the file)

**Comment by oowekyala** [comment]
The next few steps (PRs) I plan for now:
* #622
* Index generator PR. Once I'm done testing it I can submit it alongside #622

After that, we'll need to create some infrastructure around ruledef files, and rule references. I think I'll also make a new RuleSetBuilder, which instead of taking rules, takes rule references. Rules will be loaded from the ruledef files all at once. Currently, each rule reference of the form `java-basic/Rule` is parsed that way: 
  1. We reparse and recreate the `rulesets/java/basic.xml` ruleset entirely
  2. We get the rule by name ([source](https://github.com/pmd/pmd/blob/4272674488cdafbffba06e4bcab06547be7515de/pmd-core/src/main/java/net/sourceforge/pmd/RuleSetFactory.java#L315-L316))

  That looks horrendously inefficient, but could be corrected using a builder pattern and caching the XML representation of ruledef files. Using a single index for all rules will also reduce the number of different files we have to open and parse.

**Comment by jsotuyod** [comment]
@oowekyala thanks for pushing this forward. I'm sorry I couldn't sit down and read through your proposal earlier.

I'm not completely sure I understand how this will end up looking given your explanation. Could you please provide more details? In particular, I'm concerned about:

1. What is exactly a category under this definition?
1. What should these "primary categories" be? how are they different from what we currently have?
1. How do we differentiate a primary from a secondary category? Can a rule have several secondary categories?
1. How does the proposal works background-compatibly with custom plugins with rules? Those are still using old-fashioned rulesets with embedded definitions, there is no reworking into rule-defs / index generation.
1. I'm not certain why we actually need this index generation step... maybe I'm missing a point here.
1. How does the unique name per language work with custom plugins? A plugin working today may start working tomorrow simply because we officially provided a new rule that happens to have the same name. I've seen this happen on tools such as FindBugs. I think namespaces are a feature we should have.

I'll probably have plenty more questions once I get a better picture of this.

**Comment by oowekyala** [comment]
> What is exactly a category under this definition?

 A category is just a tag used to categorise rules for documentation and to ease integration into external tools. To quote from what I had written in #450:
  * Categories classify rules according to the class of problems they tackle, e.g. "code style", "control flow", "error handling", "inheritance issues", "probable bugs", "performance"...
  * Rulesets define a subset of all rules to execute, for a specific use case, e.g. "default config", "introducing pmd", "migrating junit3 to 4" + custom rulesets

The primary category, in this system, is simply a category we choose to name the folders/ packages the rule's files will live in.

> What should these "primary categories" be? 

If you're asking for a list, there's one draft in #450 

> how are they different from what we currently have?

Primary categories describe the rule, such that "basic" or "controversial" will not be primary categories. 

> How do we differentiate a primary from a secondary category?

We have to settle over a finite set of primary categories for each language that has enough rules to warrant the creation of several categories. Primary categories determine which files a rule lives in, so if we don't want it to move often we should choose some category that we're confident will not change. Though ultimately the choice is arbitrary

> Can a rule have several secondary categories?

Yes

> How does the proposal works background-compatibly with custom plugins with rules?

What plugins are you referring to? Do you mean they have a ruleset file with *custom* rules embedded into the plugin? Or that they have a copy of PMD's internal ruleset files? I've not thought about this tbh

> I'm not certain why we actually need this index generation step

Yeah we wouldn't need it if we used a single xml file for all rules (can be done too). In this proposal, I put forward index generation so that we may have multiple small files to declare rules, instead of a huge file within which the only way to navigate is ctrl+f. So we work on smaller files, with rules grouped thematically, then PMD actually only uses the generated mega-file when building rulesets. 

> How does the unique name per language work with custom plugins?

The unique name per language is not a strict constraint on rules. Several rules could have the same name in the same ruleset. Enforcing unicity of names within a language is just a way to ensure that a reference such as `java/EmptyWhileLoop` refers unambiguously to *one* rule among those that are shipped with PMD.

But I guess same question here: what plugins are you referring to? I'll check their code in the weekend

**Comment by jsotuyod** [comment]
@oowekyala sorry for the delay.

> But I guess same question here: what plugins are you referring to? I'll check their code in the weekend

Any organization can write a custom Java project with a custom ruleset and define rules, both in XPath and in Java. Then, by simply adding such jar to the PMD classpath (specially easy under Maven / Gradle), have them accessible at runtime. Such plugins may or may not be open source, we can't simply check their code. On an extreme case, the rule code can be along side the code they want to analyze (see [this integration test form Gradle](https://github.com/gradle/gradle/blob/master/subprojects/code-quality/src/integTest/groovy/org/gradle/api/plugins/quality/PmdPluginAuxclasspathIntegrationTest.groovy), where they actually write an ad-hoc Java rule to check if PMD is getting values passed properly)

We need to provide a solution that we are certain will work regardless of whatever they may have done.

Under such constraint, we can't assume unique rule names per language. Also, there would be no index generation for such rules, and even if it was (requiring rework from users!), it would be separate from the one we build and ship in PMD, being outside the "standard" index for the language.

Up to this day I don't see how this whole index thing works, and I have great doubts it will work along third-party rules.

What I think we can do and makes sense is reworking the current rulesets into categories, and have rulesets reference them either completely or by individual rules (as we currently do, *lang-category/rule*). We **may** introduce a new schema version and split `<rule>` from `<rule-ref>`, but that's an extra step that requires as to split parsing... In doing so, we may want to look at [how codenarc dealt with it](http://codenarc.sourceforge.net/codenarc-creating-ruleset.html#Creating_an_XML_RuleSet_File), as it's clearly inspired on PMD, but did a similar split.

I feel the discussion of such categories is something we need to start soon. We could go with an approach like FindBugs (that is, type of bugs: Correctness, Bad Practice, Malicious Code, Multithread, Performance, Security, Dodgy Code); by topic (testing, logging, performance, OO-design, etc.); or maybe something completely different.

**Comment by adangel** [comment]
@oowekyala @jsotuyod I've now also found the time to read through this :)

**rule definition and rulesets**
I like the idea, to define the rules separately from the rulesets. This would be just a convention, we do for PMD, e.g. we might have a single ruleset.xml per rule, which just defines one rule (by convention). Then we have the official built-in rulesets, which consist purely of rule references. It feels right to separate them, because we either define a rule or use a rule. And I think, it gives us more flexibility.
Since we need to keep backwards compatibility (at least, we should try), we wouldn't change the schema at first.

Example:
* `pmd-java/src/main/resources/rules/pmd/java/JumbledIncrementer.xml`: That would be a ruleset, that just defines the single rule JumbledIncrementer. Namespace would be "pmd", language is "java".
* `pmd-java/src/main/resources/rulesets/java/basic.xml`: The current ruleset file - we need to keep it. It will contain a rule-ref: `<rule ref="rules/pmd/java/JumbledIncrementer.xml/JumbledIncrementer" />`. We can provide a shortcut for this (similar what we do currently with java-$ruleset). It could look like this `<rule ref="pmd:java/JumbledIncrementer"/>`. This would require a code change.
* We need to make sure, that rule references can be chained, e.g. a custom-ruleset.xml might contain this reference right now: `<rule ref="java-basic/JumbledIncrementer"/>` which is expanded to "rulesets/java/basic.xml/JumbledIncrementer" which in turn will be expanded to "rules/pmd/java/JumbledIncrementer.xml/JumbledIncrementer".

Later, we could enforce the split rule definition / rulesets. We could even think about splitting the maven module "pmd-java" into "pmd-java-engine" and "pmd-java-rules", providing a pattern to extend PMD with custom rules.

**Unique rule name per language**
Honestly, I don't know, what actually happens currently, if we had two different rules with the same name in a single ruleset. It looks like, we would just add the rule to the list (since it is not a reference, we don't call addRuleReplaceIfExist...) . But e.g. the method [getRuleByName()](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/RuleSet.java#L421) has a big problem already now. I think, I once adjusted the eclipse plugin to deal with rulename+language, assuming the "unique rule name per language" holds true.
I guess, at a very least, we should add a warning when a ruleset contains two rules with the same name and same language. It looks dangerous.

**Namespaces**
Regarding the custom rules, that users might use, the idea of having namespaces sounds promising.
As in the example above, the namespace could be just another directory or for the simplified rule-ref-syntax a prefix.
But thinking about the unique rule name problem, what I'd like to have is some kind of unique identifier of a rule. E.g. if you have a rule instance, you could ask: "Who are you?" And the instance should be able to say: "I'm rule XYZ for language ABC in namespace PMD". In our own namespace, we can ensure to have unique rule names (once we cleaned up the typeresolution ruleset). The rule could have a method `getNamespace()`.
Question: Do we explicitly set the namespace or fill it by convention when we load the rule? By convention would make it easier to handle for us, the ruleset(builder) would automatically set the namespace attribute based on the file path...

**Categories**
I would take the same/similar categories other tools use. It makes it easier later when generating reports:

Findbugs: Bad practice, Correctness, Experimental, Internationalization, Malicious code vulnerability, Multithreaded correctness, Performance, Security, Dodgy code
[Codacy](https://support.codacy.com/hc/en-us/articles/207280469-What-do-the-different-issues-categories-mean-): Error prone, Code complexity, Code Style, Unused Code, Security, Compatibility, Documentation, Performance
[CodeClimate](https://github.com/codeclimate/spec/blob/master/SPEC.md#categories): Bug Risk, Clarity, Compatibility, Complexity, Duplication, Performance, Security, Style

With the above file structure in mind, we could provide an alternative rulesets, that collect the built-in rules as categories, e.g.
`pmd-java/src/main/resources/category/java/dodgycode.xml` which we could make available as `<rule ref="pmd:java-dodgycode">` (Note: namespace is pmd, without namespace, it would refer to the old style src/main/resources/rulesets/... path).


**Other thoughts:**
* We currently have the file [ruleset.properties](https://github.com/pmd/pmd/blob/master/pmd-java/src/main/resources/rulesets/java/rulesets.properties) which lists all built-in rulesets. This is used at least by the eclipse-plugin. And I also used it to generate the rule documentation pages. This part will need to be adjusted.

* We have also support for writing rulesets in PMD: [RuleSetWriter](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/RuleSetWriter.java). We should adjust this, too.

* When specifying the short version of the rule refs (e.g. "pmd:java-dogycode/JumbledIncrementer") we could take "#" as a separator, especially for selecting a single rule, like "pmd:java-dodgycode#JumbledIncrementer". However, this implies a unique name within the category at least. (The name would act like an xml ID).

**Comment by jsotuyod** [comment]
I like the idea of having things defined by convention rather than enforcing it... however, I believe we aim to:
1. have each rule in exactly one category
2. have all rules under a category

So, do we really need to split the rule definition from it's "categorization", having the risk of failing at either of both targets?

I'm thinking we may just use the current approach, just tidy up the categories, and maybe move them to a proper location such as `pmd-java/src/main/resources/rules/pmd/java/performance.xml`.

On top of that we may ship rulesets with different picks (`rulesets/pmd/java/all.xml`, `rulesets/pmd/java/android.xml`, `rulesets/pmd/java/j2ee.xml`, etc.) and let users create their own.

Without changing the schema, a rule *can* be defined in a ruleset (allowing backwards compatibility for custom rules), but we simply won't do it by convention.

For backwards compatibility, we may simply recreate the current rulesets with their current rules referenced, and mark them as deprecated someway to log proper warnings inviting people to look into the new ones (specially since Maven / Gradle have very basic defaults that don't fully tap on PMD's power).

I support @adangel's ideas on namespacing and unique names... I'm not sure how we would provide a default implementation that would work for non-pmd rules 'though. Nor if we will be able to do it without a breaking API change. If we can't, I'd probably drop it, but otherwise I support it.

**Comment by oowekyala** [comment]
> Any organization can write a custom Java project with a custom ruleset and define rules, both in XPath and in Java. 

@jsotuyod Ok I was confused by the term "plugin", thought you meant things like the maven plugin. So just to clarify on my original proposition:

Third-party rules (plugins) would *not* impacted by my proposition. We only change the way we treat rule elements with a `ref` attribute. To build a rule, we need a `rule` xml element. Then
* If it uses a `ref`, then the referenced rule is defined within PMD (IIRC one cannot reference a custom rule from custom ruleset to ruleset). Thus we need to find the rule definition element within our internal files. Currently we use the ruleset part of the ref to identify the file, but if we dump every rule of a language within one huge file, then we *know* the rule is defined in that one file. It becomes very easy to find the definition (just find the element with the correct `name` attribute), and instantiate the rule (using a RuleFactory). Old refs like `java-basic/Foo` can be converted to `java/Foo` on the fly, so old references would still work too
* If it doesn't use a `ref` (third-party rule), then the definition of the rule is the current element. That's a no brainer, we just use a RuleFactory to instantiate it. 

So in my proposal, the "unique rule names per language" is just a convention so that the index doesn't end up having several rules with the same name. That's because it's *indexed on the rule names*. Other than that, we can have two rules with the same name in the same ruleset (current definition of rule equality compares name, class, and properties IIRC), or at least my proposal doesn't change the current behaviour on this part.

> Without changing the schema, a rule can be defined in a ruleset (allowing backwards compatibility for custom rules), but we simply won't do it by convention.

Yup

> So, do we really need to split the rule definition from it's "categorization"

I totally agree we shouldn't, that's why I proposed the thing about "primary categories", though maybe that's not the best approach. Either way, generating a single file from all the files we use allows us to organise our rules however we may like, eg with one file per rule like @adangel suggested, or just in their categories like you and I suggest Juan. In the end all rules end up in the same file.

@adangel I think namespaces are a good idea too. Though like Juan I'm not sure how you would implement it right away. Also, I'm not sure it's that useful yet since users can't reference their own custom rules (can they?), but namespaces could help the implementation of that feature


> With the above file structure in mind, we could provide an alternative rulesets, that collect the built-in rules as categories, e.g. pmd-java/src/main/resources/category/java/dodgycode.xml which we could make available as `<rule ref="pmd:java-dodgycode">`

I'm not sure how you would differentiate a category reference from a ruleset reference then... how would you reference the ruleset java-basic?

**Comment by adangel** [comment]
> IIRC one cannot reference a custom rule from custom ruleset to ruleset

In theory, it should be possible - not sure however, how much this (unknown/undocumented) feature is used: the ref attribute value is current interpreted that way: If it ends with  ".xml", then it's referencing a complete ruleset on the classpath. So, as long as the custom rulesets are available on the classpath, we don't restrict referencing them on other custom rulesets. Similar, if it uses the pattern "<ruleset>/<rulename>", then the ruleset part is interpreted as above (the ruleset can be any file on the classpath). If the ruleset part doesn't end with ".xml" and is of the pattern "<language>-<rulesetname>", then try to find the file on the classpath called "/rulesets/<language>/<rulesetname>.xml". It just a transparent conversion, similar like your suggestion `java-basic/Foo` -> `java/Foo`.

Whether we should allow such a scenario? I think, we should - otherwise we would artificially limit the functionality. There could be a big company, that maintains shared code quality rules for all company wide projects. They could provide a company-wide ruleset which could be references by a project-specific ruleset... Not sure, if this scenario is actually reality :)



>> So, do we really need to split the rule definition from it's "categorization"
>
> I totally agree we shouldn't, that's why I proposed the thing about "primary categories", though maybe that's not the best approach. Either way, generating a single file from all the files we use allows us to organise our rules however we may like, eg with one file per rule like @adangel suggested, or just in their categories like you and I suggest Juan. In the end all rules end up in the same file.

Technically it's not necessary and I agree, it is much effort with only little to gain - since it would be anyway just a convention, rather than enforced. The separation sounded like a good idea to me, but to make more sense of it, we probably would need to change the way, how rules are defined (e.g. we could add a primary category "attribute" and generate the category view from all the rules without explicitly needing to create/maintain these primary category rulesets, splitting the schema into rule-def and rule). But aiming for 6.0.0, we probably shouldn't do any such big changes.

I think, the naming that @oowekyala defined, is good: a `rule`-tag without a `ref` attribute is a **rule-definition** and a `rule`-tag with a `ref` attribute is a **rule-reference**. What PMD needs to have, is an overview of all the defined rules (whether we scan the classpath somehow on the fly or use a rule index, that's implementation detail). With that knowledge, we should be able, to uniquely identify rules: For a given rule, it would be the ruleset (or more exactly the path to the ruleset file in the classpath), in which it has been defined + the rule name. We can also do a consistency check, that no rule class is used in more than one rule definition. This would be an error (probably caused by copy-paste).

> I think namespaces are a good idea too. Though like Juan I'm not sure how you would implement it right away. Also, I'm not sure it's that useful yet since users can't reference their own custom rules (can they?), but namespaces could help the implementation of that feature

The namespace implementation, I had in mind, would be very, very simplistic: We can define the rule-reference-string as we like. Currently supported formats are:

* `/some/path/file.xml/RuleName` -> single rule
* `/some/path/file.xml` -> whole ruleset
* `java-basic/RuleName` -> java-basic will be expanded to: `/rulesets/java/basic.xml` -> single rule
* `java-basic` -> java-basic will be expanded to: `/rulesets/java/basic.xml` -> whole ruleset

We keep all of the above as valid, but add these ones:

* `namespace:java-custom/MyRule` -> since it starts with a string followed by a colon, it's a namespace. The whole ref will be expanded to `/rules/namespace/java/custom.xml` and we reference in that ruleset the rule "MyRule".
* `namespace:java-custom` -> whole ruleset, expanded version `/rules/namespace/java/custom.xml`.

We would define, reserve and claim the namespace "pmd" for us, to be used for the built-in PMD rules.
I imagine, a Rule instance knows at runtime its namespace, and thus can generate a rule reference string to itself (just a little sugar I think would be useful for the eclipse-plugin or any future ruleset editor tooling).
The namespace wouldn't be set explicitly, but only implicitly by defining the rule in a ruleset within a specific folder structure.
The downside is, the namespace usage is not enforced and anybody could define a ruleset in this namespace (it's just a location in the classpath...).
If a ruleset is placed under `/com/example/quality/pmd/rules.xml`, then all rules defined (if any) will have no namespace and no short-variant of the reference, but only the full one, consisting of the full path.

> I'm not sure how you would differentiate a category reference from a ruleset reference then... how would you reference the ruleset java-basic?

The ruleset java-basic would be references like now, as `java-basic` or `/rulesets/java/basic.xml` - to be backwards compatible. But you're right: My system would not differ between a "category ruleset" and a "miscellaneous/custom ruleset". It would be again just convention: We as PMD provide under the pmd-namespace only rulesets, which group the rules into their primary categories.


> I'm thinking we may just use the current approach, just tidy up the categories, and maybe move them to a proper location such as pmd-java/src/main/resources/rules/pmd/java/performance.xml.

Exactly, that would be the first step. I would suggest, to make this ruleset available as `pmd:java-performance`.

> On top of that we may ship rulesets with different picks (rulesets/pmd/java/all.xml, rulesets/pmd/java/android.xml, rulesets/pmd/java/j2ee.xml, etc.) and let users create their own.

Right. We might just make them available under their full path, or we could provide a short-cut reference, e.g. `pmd:java-all`. In that case, we would first search for a "category" ruleset under `rules/pmd/java/all.xml` and if none found, fall back to search under `rulesets/pmd/java/all.xml`.

**Summary:**
I'll try to summarize, where we are (please correct me, if I'm wrong):

* The plan is, to have a primary category for each rule
* The primary category is implemented as a ruleset
    * First task: Decide on the primary categories
    * Second task: Move the rules into the corresponding category and add a reference in the "legacy" ruleset
* Rule names are unique within a language and a primary category (within our built-in rulesets)
* No change to the ruleset schema for now. No split of the rule definitions from its primary category.

* Namespaces:
   * Suggestion is: implement as a subdirectory encoded in a rule-reference string
   * Would be 3rd task: Implement/Adjust rule-reference parsing
* Provide user-oriented, ready-to-use, use-case-specific rulesets: all.xml, start.xml, android.xml, migrating.xml, ...
* Where are we on introducing secondary categories/tags/etc? Do we still need it?

**Comment by oowekyala** [comment]
@jsotuyod @adangel Sorry for the delay

> Whether we should allow such a scenario? I think, we should - otherwise we would artificially limit the functionality. 

Ok, that makes sense

> namespace:java-custom -> whole ruleset, expanded version /rules/namespace/java/custom.xml.

> We would define, reserve and claim the namespace "pmd" for us, to be used for the built-in PMD rules.

> We as PMD provide under the pmd-namespace only rulesets, which group the rules into their primary categories.

@adangel So if I understand correctly, `pmd:java-xxx` would refer to a "category ruleset" named xxx, living at `/rules/pmd/java/xxx.xml`, while `java-xxx` would refer to a ruleset at  `/rulesets/pmd/java/xxx.xml`? That looks confusing to me. I don't think it's a good idea to lengthen rule references even more... Plus, adding ways to reference the same rule (one through the ruleset, the other through the category) looks like moving the problem around. Users are still coupled to the file the rule is defined in 

If the objective is to make any rule capable of identifying itself at runtime, we could simply use the classpath url of its definition file as an (implicit) namespace. That removes the shortcoming of rules having no namespace if they're not defined in a conventional path.

Since the path could be quite long, it would be convenient to define short aliases that could be used as namespace prefixes. We could use a ruleset-wide "property section" where we define ruleset-wide config. I think this has already been suggested for other purposes but can't find it. We could define eg an alias for a ruleset path as a namespace prefix. E.g.
```xml
<ruleset>
<properties>
  <namespace-alias ns="company" path="/com/example/quality/pmd/rules.xml" />
</properties>
```
which would allow people to references the rules defined in that file as `company:RuleName`. We keep a wide degree of flexibility regarding the location of rule definition files, but rules know their location and can create a reference to themselves at runtime. 

This prefix is different from your namespace because it identifies a single file, while your namespace prefix identifies a directory. Under that scheme, and assuming we use a single file per language to declare rules (generated or not), we could reserve all prefixes that begin with "pmd" and access rules using eg `pmd-java:EmptyWhileLoop`, where "pmd-java" is one such alias, expanded to "/rules/pmd/java/all-rules.xml" (assuming that's the location of our single file). We could reference rulesets that way too with a bit more work

That way, references to standard rules don't depend directly on the specific location of the file. References to custom rules are cleaner looking and more maintainable (the path to the ruleset is mentioned only once). And references get shorter instead of longer. Though, in your summary, you make no mention of references like "java/EmptyWhileLoop"; is that a lapse, or are you putting the suggestion aside? 
Cheers

**Comment by adangel** [comment]
@oowekyala 
> So if I understand correctly, pmd:java-xxx would refer to a "category ruleset" named xxx, living at /rules
> /pmd/java/xxx.xml, while java-xxx would refer to a ruleset at /rulesets/pmd/java/xxx.xml?

Correct - it's for keeping backwards compatibiliy.

> I don't think it's a good idea to lengthen rule references even more...

Agreed.

> That way, references to standard rules don't depend directly on the specific location of the file. 

Yes, I think that's the key point we need to achieve.


@oowekyala @jsotuyod 
I've created this wiki page to describe a proposal: https://github.com/pmd/pmd/wiki/Rule-Categories

* First part [Proposed Categories](https://github.com/pmd/pmd/wiki/Rule-Categories#proposed-pmd-categories): What are the categories, we want to group the rules in? I've tried to make a suggestion... But it feels like still too many categories, I came up with 9 categories now.

* Second part: [Using Rules](https://github.com/pmd/pmd/wiki/Rule-Categories#using-the-rules-in-custom-rulesets): I've not added namespaces, not sure if we actually need it for the first shot (only mentioning Clément's `pmd-java:JumbledIncrementer` example). I think, it would add unnecessary complexity, if we need to explain, how to set/use namespaces for custom rulesets.
Having said that, maybe we should prevent points 3 and 4 completely (referencing via category, referencing a whole category)?

I think, regarding the categories, we need to answer this question:
* Is the use case for categories, to reference a whole category in a custom ruleset?
* Or is the use case for categories just documentation - helping the users to find/discover rules - and later on reporting - categorize the violation?

Cheers,
Andreas

**Comment by oowekyala** [comment]
@adangel 
> I've created this wiki page to describe a proposal: https://github.com/pmd/pmd/wiki/Rule-Categories

That's amazing! Imo the number of categories is ok, less than that would bloat categories, more than that would complicate things. Mapping these categories to those of other tools is not trivial though, it may not be possible to find a common denominator... The easiest solution for those tools would perhaps be to maintain one custom ruleset for each of their categories, but we could let them manage that themselves



> Or is the use case for categories just documentation - helping the users to find/discover rules - and later on reporting - categorize the violation?

I'd say, considering categories that way would make life easier. Referencing a full category may contain inconsistencies, eg contradictory rules (CallSuperFirst vs CallSuperLast). Big categories may contain a lot of undesired rules too, though we could argue the same of rulesets. 

Besides, I think categories ought to really be separate from rulesets, ie unreachable with a reference like `rulesets/java/correctness.xml` or `java-correctness`. It's perhaps just dogmatic of me, but that way, there's nothing "special" about our rulesets, which means they're better learning resources for users, and the directory `rulesets` is self-documenting. It may be easier to explain the difference to users, and to think about them as different things, if we enforce a conceptual separation in our codebase. 
 
For that reason I still think category files should go into a separate file hierarchy ("rule", "ruledef", "rules" or whatever). That would prevent moving rulesets into another "tags" directory as well, and make the current rule reference syntax sufficient to cope with the new changes organisation (because no need to introduce new syntax to refer to categories). I feel like `/rulesets/java/tags/<tagname>.xml` feels more of a hack and would need special syntax to shorten it

Having said that, even in the situation I describe, full category reference would not be hard to implement, assuming the categories are mapped one-to-one to xml files. Though I'd tend to write it differently than ruleset reference, eg `java#correctness` or similar. Maybe even use a special xml element. They'd be more decoupled from rulesets.

Cheers

**Comment by jsotuyod** [comment]
> We can also do a consistency check, that no rule class is used in more than one rule definition. This would be an error (probably caused by copy-paste).

Big exception on XPath rules though.


> * First part Proposed Categories: What are the categories, we want to group the rules in? I've tried to make a suggestion... But it feels like still too many categories, I came up with 9 categories now.

The current list seems ok. I just have 2 suggestions:

1. I'd add a `security` category. We have no Java rules for it *at the moment* (we may want to add some checks on SQLi with JDBC for instance), but we do for Apex, and I believe it's better to keep them apart to enhance their criticality.
2. I'd consider merging `Complexity / Code Size` into `Design / API`. After all, all complexity / code size issues are related to problems in the design / API / modularization.

> * Second part: Using Rules: I've not added namespaces, not sure if we actually need it for the first shot (only mentioning Clément's pmd-java:JumbledIncrementer example). I think, it would add unnecessary complexity, if we need to explain, how to set/use namespaces for custom rulesets.
Having said that, maybe we should prevent points 3 and 4 completely (referencing via category, referencing a whole category)?

Agreed, I'd drop namespaces for 6.0.0, it's necessity is still unclear to me.

I'd personally also drop the `<rule ref="java/JumbledIncrementer" />` notation. I believe we already have too many ways to reference a rule to keep adding new ones, and I see little value with it. It's only a few chars shorter, and "detaching" the rule from the category for future changes, which shouldn't be an issue since we are attempting to bring consistency to them; on the other hand it adds the extra complexity of having to deal with this reference format, and making it harder on users that have to deal with yet another way to reference rules. Plus, this format would only work for internal rules, making it impossible for us to deprecate / remove the other formats in favor of this one, confusing users when writing a ruleset.

As for the index, I just realized that it would be an exact match for the `all.xml` ruleset, which should be autogenerated to avoid missing them out. Even if we don't use it as an index itself.

> I think, regarding the categories, we need to answer this question:
> * Is the use case for categories, to reference a whole category in a custom ruleset?
> * Or is the use case for categories just documentation - helping the users to find/discover rules - and later on reporting - categorize the violation?

Since categories are basically rulesets, and we are doing this by convention, I don't think we should enforce such a prohibition. It would be arbitrary, confusing for users, and mostly pointless. Given we provide good presets and proper documentation on how to build a ruleset, I don't see much of an issue on this as to bother building the logic to prevent such usage.

**Comment by jsotuyod** [comment]
Upon further review of the proposed category list, I'd also consider merging Correctness into Error Prone. Most of the rules there end up detecting things that either:

* don't work at all as expected (ie: EqualsNull)
* are confusing / error prone (ie: NonCaseLabelInSwitchStatement)
* are actual issues (ie: CloseResource)
* may break depending on how you use the code (ie: CallSuperFirst / CallSuperLast)

Therefore, I'd stick to just 8 categories:

1. Error Prone
2. Code Style
3. Multithreading
4. Best Practices
5. Performance
6. Documentation
7. Design
8. Security

We may reduce this to 7 if we decide to merge Multithreading into Error Prone too, given not all languages have multithread support.

I suggest that once we agree on the categories, we start a separate issue to discuss any further changes to on which category rules belong taking @adangel's proposal as a starting point. That way we can keep this conversation focused on how to implement categories / ruleset changes / rule referencing.

**Comment by adangel** [comment]
> I'd say, considering categories that way would make life easier. Referencing a full category may contain inconsistencies, eg contradictory rules (CallSuperFirst vs CallSuperLast). Big categories may contain a lot of undesired rules too, though we could argue the same of rulesets.

I believe, the main use case for categories is not, to select all rules of a category - but to understand the impact of the issues. Although we probably won't forbid to just select a complete category/ruleset.
And you are right, we already have the same issues with rulesets - that's why we propose, to create customized rulesets with exactly the rules enabled, that make sense for the project at hand.

> For that reason I still think category files should go into a separate file hierarchy 

I tend to agree with you. I think, for me the reason is, to have it separated to say: "Look, this is where the new category rulesets are - all the other ruleset files in the old place will be deprecated" - so clearly separate the old way from new way.

> Agreed, I'd drop namespaces for 6.0.0, it's necessity is still unclear to me.

That would be a way, to select the "new way". But I think, I understand now, that this separation is not a technical issue, it's just "something else" :) Assuming, we deprecate all rule references, that are contained in the old rulesets and eventually remove the old rulesets with 7.0.0, the end result is the same: Only the category rulesets are left.

Oh, one challenge I see here: E.g. consider the ruleset design.xml - with the proposed categories, we will have a filename clash. We would add a couple of rules here from other rulesets and deprecate some rules, that we moved to other categories. Result is, that when someone uses the complete design.xml ruleset, more rules will be enabled...  but maybe, that's just an edge case - and as I mentioned earlier - referencing a complete ruleset is not what PMD is suggesting anyway...

I'll update the wiki page with the 8 categories and give it a try with a PR today...

**Comment by jsotuyod** [comment]
> Oh, one challenge I see here: E.g. consider the ruleset design.xml - with the proposed categories, we will have a filename clash.

Not if we use a separate file hierarchy. `pmd-java/src/main/resources/category/java/design.xml` would not clash with `pmd-java/src/main/resources/rulesets/java/design.xml`. The old `java-design` reference would still point to the ruleset.

**Comment by adangel** [comment]
> I thought we had agreed to keep categories on a different directory, such as pmd-java/src/main/resources/category/java/codestyle.xml...

Since we don't introduce a new rule-ref shortcut (is that correct?), I thought, we'd also skip the directories....
If we create a separate hierarchy, let's say `category`, then the only way to reference these category rulesets is via a full path:

    <rule ref="category/java/design.xml/AbstractClassWithoutAnyMethod" />

There will be no shortcut additional shortcut. If someone still uses `java-design/AbstractClassWithoutAnyMethod`, this will always be resolved only to `rulesets/java/design.xml/AbstractClassWithoutAnyMethod`. (Ok, and there we have a reference pointing to the category for backwards compatibility - which hopefully produces a warning "rule has been moved ...").

Hm...it seems we do not document the short form currently ([Making Rulesets](https://pmd.github.io/pmd/pmd_devdocs_making_rulesets.html)). Not too bad....

So, in summary, we'll suggest to the end-users, when they create their own custom rulesets, to select rules like this:

    <rule ref="category/java/design.xml/AbstractClassWithoutAnyMethod" />
    <rule ref="category/java/design.xml/AvoidCatchingGenericException" />
    <rule ref="category/java/design.xml/AvoidDeeplyNestedIfStmts" />

Correct?

**Comment by jsotuyod** [comment]
@adangel that was my understanding so far.

The shortcuts are good for running PMD (since those are the presets for quick start), but having to use full-paths for custom rulesets doesn't seem bad to me. We may add a shortcut later on as we deprecate some of the older reference formats / add namespaces, but it doesn't seem so bad at this point.
